lib/AstOfLlbc.ml

```diff
@@ -922,70 +943,140 @@ let rec build_trait_clause_mapping env ?depth (trait_clauses : C.trait_clause li
         (* 1. Associated constants *)
         List.map
           (fun (item_name, typ) ->
-            (C.Clause (Free clause_id), item_name), ClauseConstant (trait_decl.C.item_meta.name, typ))
+            let trait_name = trait_decl.C.item_meta.name in
+            let pretty_name = string_of_name env trait_name ^ "_" ^ item_name in
+            let t = substitute_visitor#visit_ty subst typ in
+            let t = typ_of_ty env t in
+            TraitClauseConstant { item_name; pretty_name; clause_id = clause_ref }, t)
           trait_decl.C.consts
         (* 2. Trait methods *)
         @ List.map
             (fun (item_name, bound_fn) ->
-              let fun_decl_id = bound_fn.C.binder_value.C.fun_id in
-              let decl = env.get_nth_function fun_decl_id in
-              let ts =
-                {
-                  K.n = List.length trait_decl.generics.types;
-                  n_cgs = List.length trait_decl.generics.const_generics;
-                }
+              let trait_name = trait_decl.C.item_meta.name in
+              let pretty_name = string_of_name env trait_name ^ "_" ^ item_name in
+
+              (* First we must substitute the trait generics into the method
+                 reference. *)
+              let bound_fn : C.fun_decl_ref C.binder =
+                substitute_visitor#visit_binder substitute_visitor#visit_fun_decl_ref subst bound_fn
+              in
+              (* From this we construct a `fun_sig binder` which binds the
+                 method-specific generics. *)
+              let bound_sig : C.fun_sig C.binder =
+                let method_decl_id = bound_fn.C.binder_value.C.fun_id in
+                let method_decl = env.get_nth_function method_decl_id in
+                (* Get the signature from the `fun_decl` and substitute it to
+                   be valid under the binder. *)
+                let signature =
+                  substitute_visitor#visit_fun_sig
+                    (Charon.Substitute.make_subst_from_generics method_decl.signature.generics
+                       bound_fn.binder_value.fun_generics)
+                    method_decl.C.signature
+                in
+                { C.binder_params = bound_fn.binder_params; binder_value = signature }
+              in
+
+              (* We then construct a polymorphic signature for this method.
+                 Its generics are the method generics (the ones in the binder).
+                 *)
+              let method_sig =
+                Charon.Substitute.(
+                  (* Variables bound in the inner binder are `Bound`, which
```

Comment:
can you say more about what is not handled here, and where it should be handled?